### PR TITLE
Added script files support

### PR DESCRIPTION
**Pull request by peterdeme**
# Pull Request Check List

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

# Feature: script-files section in pyproject.toml

### Reasoning

This one is a long overdue, see: https://github.com/sdispater/poetry/issues/241 (June 2018)

Basically, our company wants to migrate over Poetry but it's impossible since Poetry does not provide a way to add non-Python scripts to the dist package and our migration project is pretty much blocked 😞. This is a very basic feature of setup.py See here: [docs](https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-scripts-keyword-argument).

Note: this is **not the same as** `entry_point`! This setuptools feature basically just copies over the given files to the virtualenv's `bin` folder so that it'll be available on the `$PATH` and you can simply invoke it.

### Example

```toml
[tools.poetry]
package = "my_package"
script-files = ["bin/run_db_migrations.sh"]
```

Then:

```bash
poetry install
poetry run sh run_db_migrations.sh
```

### Testing

- I added a couple of automated tests
- I tested this **manually** with one of our repositories and it worked

### Possible discussion

Naming. Well, I was scratching my head to come up with a decent name. `script-files` sounds sensible to me, but I can change it if you guys want.

edit.: for some reason the GitHub Actions build on Windows failed for me. Looks like it tried to run the commands in powershell instead of cmd - I don't know why, maybe GitHub changed something this weekend? Anyway, I enforced `cmd` in the build declaration, the build looks good now.

### CODE REVIEW

**Comment by peterdeme** [comment]
Can this have some visibility?

**Comment by peterdeme** [comment]
@stephsamson - what do you think?

**Comment by jleclanche** [comment]
I'm not a fan of this approach… There is solid prior art on how to do this in yarn/npm and pipenv.

This is completely intuitive:

```toml
[tool.poetry.scripts]
dosomething = "./path/to/dosomething --arg"  # Path relative to pyproject.toml
```

I also agree with your assessment in [this comment](https://github.com/sdispater/poetry/issues/241#issuecomment-546045968) that the "scripts" naming for what "scripts" currently is doesn't make sense if it's just to replicate setuptools' entrypoints.

It should either be renamed tool.poetry.entry_points, or tool.poetry.scripts should be able to safely detect whether you're attempting to run a script vs. an entrypoint.

**Comment by jleclanche** [comment]
Just to clarify the comment above, this is more in response to this PR resolving #241, which I don't think it does; IMO that issue has more to do with poetry's usage as a project manager than a project packager. Being able to include scripts files in the resulting package makes sense, but isn't the whole picture.

**Comment by peterdeme** [comment]
@jleclanche yeah, I was also confused about the terminology.
I think it's a bad idea to introduce such a breaking change so I did not rename existing functions.

`Being able to include scripts files in the resulting package makes sense, but isn't the whole picture.`

Care to elaborate?

**Comment by peterdeme** [comment]
@sdispater opinion?

**Comment by elihunter173** [comment]
To avoid changing `scripts` to `entry_point` and breaking a bunch of people's projects (although I agree entrypoints makes more sense), I think calling the new functionality `tasks` is intuitive.

For example, if I have something like below in my pyproject.toml, then I'd expect to be able to do either `poetry dosomething` or `poetry task dosomething` to execute the task.

```toml
[tool.poetry.tasks]
dosomething = "./path/to/dosomething --arg"  # Path relative to pyproject.toml
```

**Comment by peterdeme** [comment]
I’m still a bit confused how is this relevant to this feature. This isn’t about running arbitrary scripts but copying any kind of files (not just executables) to a package.

**Comment by danbradham** [comment]
Glad to see this PR @peterdeme thank you for working on it. I'm not sure why everyone is misunderstanding this PR as well as #241, but hopefully your PR can add some clarity to the discussion. 

The scripts keyword of setup() allows you to install non-python scripts along with a python package. This should not be confused with running arbitrary development tasks.

I personally feel like tool.poetry.scripts should be renamed for the sake of clarity, but, that seems like a larger discussion / outside of the scope of this PR. The script-files keyword seems fine, I could also see something a bit more explicit like include-scripts working well.

Can @sdispater or any other maintainers weigh in on this? I think there are probably quite a few projects that are unable to use poetry due to this missing feature.

**Comment by astronouth7303** [comment]
Any motion on this? I just reworked a bunch of stuff in my project to work around #241.

**Comment by peterdeme** [comment]
@astronouth7303 the PR’s been open for 4 months. Looks like they don’t have bandwith to review it or it’s a low priority to them

**Comment by geckon** [comment]
I'd also love to see attention to this as it's something I think poetry definitely should support and I'd love to use this feature instead of trying to come up with a work-around.

**Comment by funkyfuture** [comment]
imo `poetry` shouldn't have invented its own semantics for keywords in the first place and i think the `1.0` release was premature. i don't know how the maintainers employ the versioning scheme as the last patch release broke builds, so if breaking changes are also okay for minor releases, i'd say the current `[tool.poetry.scripts]` section should be renamed to `[tool.poetry.console_scripts]` asap and the feature in this pr should handle the `[tool.poetry.scripts]` section.

**Comment by finswimmer** [comment]
>  i don't know how the maintainers employ the versioning scheme as the last patch release broke builds,

Patch releases shouldn't break anything. So if  this really happens it wasn't intended. Could you please open an issue for this or give us a linke to an existing issue?

**Comment by funkyfuture** [comment]
> Could you please open an issue for this or give us a linke to an existing issue?

that was #2107

**Comment by sdispater** [comment]
@funkyfuture 

> i don't know how the maintainers employ the versioning scheme as the last patch release broke builds

Poetry follows semver. And the latest release did not break builds on purpose, it's just that bugs happen, unintended side effects happen. Semver is not a silver bullet against unplanned regressions. While we try our best to not break anything, we are only human and sometimes mistakes happen.

That being said, the Git issue you mentioned should now be fixed in the latest release.

**Comment by peterdeme** [comment]
@sdispater @finswimmer @stephsamson this has been open for around half year. Can you guys at least tell me if this is bad or good or something? Thank you.

**Comment by stephsamson** [comment]
Hi @peterdeme,

Thank you for your contribution! This PR is a great one. Although it is not part of the roadmap afaik, I think this should be seriously considered @sdispater.

I also think naming it `tasks` instead of `script-files` makes more sense since it’s equivalent to rake/bundle/make tasks.

**Comment by astronouth7303** [comment]
THEY"RE NOT PROJECT TASKS.

This defines a new type of installable: Files that should be placed on the $PATH without modification.

**Comment by peterdeme** [comment]
@stephsamson , thanks for the feedback. it is not really equivalent with those. As astrobouth said this one just copies a bunch of files to your PATH.

https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-scripts-keyword-argument

**Comment by stephsamson** [comment]
@astronouth7303 thanks for the clarification, I was hastily scanning the comments of the PR when I was reviewing a bunch of others and was careless. I won't do this next time and be more careful but typing in caps lock is also unnecessary :)

Regarding the keyword, I think an RFC needs to be opened on whether `poetry` should be using keywords equivalent to setup.py and would mean breaking changes.

**Comment by astronouth7303** [comment]
Sorry, the issue is like 3 miles long because of this confusion.

**Comment by stephsamson** [comment]
No worries, I can understand that it's incredibly frustrating. I'll make sure to raise this particular issue once again to the core team.

File: docs/docs/pyproject.md [hunk 1]
```diff
@@ -212,6 +212,22 @@ poetry = 'poetry.console:run'
 
 Here, we will have the `poetry` script installed which will execute `console.run` in the `poetry` package.
 
+### `script-files`
+
+You can also include non-Python scripts in the distribution package.
+
+```toml
+[tool.poetry]
+script-files = ["contrib/run_db_migration.sh", "bin/healthcheck.py"]
```
Comment (stephsamson) [inline docs/docs/pyproject.md:221]: The @python-poetry/core team has decided on going forward with keeping the `[tool.poetry.scripts]` namespace and using inline tables to have more flexibility and backwards compatibility.

File: poetry/json/schemas/poetry-schema.json [hunk 2]
```diff
@@ -82,8 +82,15 @@
                     },
                     "format": {
                         "oneOf": [
-                            {"type": "string"},
-                            {"type":  "array", "items": {"type":  "string"}}
+                            {
```
Comment (stephsamson) [inline poetry/json/schemas/poetry-schema.json:85]: This schema will have to change w.r.t. my comment above.

**Comment by peterdeme** [comment]
I’ll try to find time in the next few weeks.

**Comment by peterdeme** [comment]
Today, I had time to develop this. I opened a new PR and closing this one now.

@stephsamson Can you please take a look at https://github.com/python-poetry/poetry-core/pull/40?

**Pull request closed without merging**